src/kakarot/stack.cairo

```diff
@@ -77,30 +91,110 @@ namespace Stack {
+        );
+
+        // Return Stack with updated Len
+        let popped_len = 2 * n;
+        return (
+            new model.Stack(
+            stack_word_dict_start=self.stack_word_dict_start,
+            stack_word_dict=stack_word_dict,
+            stack_16bytes_len=self.stack_16bytes_len - popped_len,
+            ),
+            new_elements,
+        );
+    }
+
+    func dict_copy{range_check_ptr}(
+        stack_word_dict: DictAccess*, stack_len: felt, n: felt, output: Uint256*
+    ) -> (stack_word_dict: DictAccess*) {
+        if (n == 0) {
+            return (stack_word_dict=stack_word_dict);
+        }
+
+        // Get Low and High of element at position N
+        let (el_high) = dict_read{dict_ptr=stack_word_dict}(stack_len - n);
+        let (el_low) = dict_read{dict_ptr=stack_word_dict}(stack_len - n + 1);
+
+        // Save Uint256 value in array
+        let n_index = n / 2 - 1;
+        assert output[n_index] = Uint256(low=el_low, high=el_high);
+
+        return dict_copy(
+            stack_word_dict=stack_word_dict, stack_len=stack_len, n=n - 2, output=output
+        );
+    }
+
+    // @notice Pop an element from the stack.
+    // @param self - The pointer to the stack.
+    // @return The new pointer to the stack.
+    // @return The popped element.
+    func pop{range_check_ptr}(self: model.Stack*) -> (new_stack: model.Stack*, element: Uint256) {
+        let stack_word_dict = self.stack_word_dict;
+        let position_zero = self.stack_16bytes_len;
+        // Check if stack will underflow
+
+        if (position_zero == 0) {
+            with_attr error_message("Kakarot: StackUnderflow") {
+                assert 1 = 0;
+            }
+        }
+
+        // Read and Copy element at position 1(first on stack)
+        let (el_high) = dict_read{dict_ptr=stack_word_dict}(position_zero - 2);
+        let (el_low) = dict_read{dict_ptr=stack_word_dict}(position_zero - 1);
+
+        // Update and return Stack
+        return (
+            new model.Stack(
+            stack_word_dict_start=self.stack_word_dict_start,
+            stack_word_dict=stack_word_dict,
+            stack_16bytes_len=self.stack_16bytes_len - 2,
+            ),
+            Uint256(low=el_low, high=el_high),
+        );
     }
 
     // @notice Return a value from the stack at a given stack index.
     // @dev stack_index is 0-based, 0 is the top of the stack.
     // @param self - The pointer to the stack.
     // @param stack_index - The index of the element to return.
     // @return The element at the given index.
-    func peek{range_check_ptr}(self: model.Stack*, stack_index: felt) -> Uint256 {
-        alloc_locals;
-        Stack.check_underflow(self=self, stack_index=stack_index);
-        let array_index = Stack.get_array_index(self=self, stack_index=stack_index);
-        return self.elements[array_index];
+    func peek{range_check_ptr}(self: model.Stack*, stack_index: felt) -> (
+        self: model.Stack*, value: Uint256
```

Comment:
I don't understand why we need to return the stack as well as it is unchanged in peel

Reply:
dict_read()s change the dict. That's how they work in Cairo.